File: src/main/java/ru/ioleksiv/telegram/bot/api/annotations/filter/primary/PreCheckoutQueryFilter.java
package ru.ioleksiv.telegram.bot.api.annotations.filter.primary;

import ru.ioleksiv.telegram.bot.api.annotations.filter.secondary.OrderInfoFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.secondary.StringFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.secondary.UserFilter;
import ru.ioleksiv.telegram.bot.core.model.AnnotationState;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Documented
public @interface PreCheckoutQueryFilter {

    UserFilter from() default @UserFilter(AnnotationState.OFF);

    StringFilter currency() default @StringFilter(AnnotationState.OFF);

    StringFilter totalAmount() default @StringFilter(AnnotationState.OFF);

    StringFilter invoicePayload() default @StringFilter(AnnotationState.OFF);

    StringFilter shippingOptionId() default @StringFilter(AnnotationState.OFF);

    OrderInfoFilter orderInfo() default @OrderInfoFilter(AnnotationState.OFF);

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/annotations/filter/composite/PreCheckoutQueryFilter.kt
package io.github.oleksivio.tl.kbot.core.annotations.filter.composite

import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.StringFilter
import io.github.oleksivio.tl.kbot.core.model.annotation.AnnotationState

@Retention
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@MustBeDocumented
annotation class PreCheckoutQueryFilter(
    val from: UserFilter = UserFilter(
        status = AnnotationState.OFF
    ),
    val validator: Array<String> = [],
    val currency: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val totalAmount: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val invoicePayload: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val shippingOptionId: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val orderInfo: OrderInfoFilter = OrderInfoFilter(
        status = AnnotationState.OFF
    )
)
